### PR TITLE
auto awesome-lint any list added to the readme in a PR

### PR DESCRIPTION
**Pull request by RayBB**
I really appreciate this project and wanted to try to give back so it's a little easier for everyone involved.
To that end, I created this PR to **automatically lint new lists** when someone opens a PR to add them.

## Benefits
  - It will save time of maintainers by allowing them to look at the error message first
  - It will provide immediate feedback to people who make a PR
  - It helps people having trouble installing `awesome-lint`
  - It is free

## Risks
  - As far as I know, there is no risk to users or maintainers unless.

## What it does
- PR will lint **NEW** lists that are requesting to be added.
- PR will **NOT** lint [this](https://github.com/sindresorhus/awesome/) repo itself. If that functionality is desired it can happen in a separate PR (see https://github.com/sindresorhus/awesome/pull/1794).
- PR will **NOT** lint the current lists in the readme.

## How it does it
This PR creates a GitHub action that will do the following every time a new commit is pushed to an open PR:
1. Find the URL that is being added using `git diff`
2. Clone that repo
3. Run `awesome-lint`on that repo.

The above 3 steps will **ONLY** run if `readme.md` is changed.
If readme is changed and no new URL is detected in the readme it will fail with a helpful error message.
Error can be ignored if it is a cosmetic PR

[Here](https://github.com/RayBB/awesome/pull/1) are examples of failing commits.

## Final notes

"Squash and Merge" this PR or let me know and I will squash it manually.

Feel free to edit this any way you see fit or ask me to make the edits. 👍
If there are any questions on how this works drop them here and I'll try my best to answer.

Again, thanks so much for all the hard work the maintainers put into this repo! 👏

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      # If there's no repo found leave an error message
+      - name: validate repo
+        run: if [ -z ${REPO_TO_LINT+x} ]; then echo "Link to your list using the form https://....#readme" && exit 1; else echo "Repo detected '$var'"; fi
+
+      # Clone the repo into a new folder
+      - name: clone
+        run: echo "Cloning $REPO_TO_LINT" && mkdir cloned && cd cloned && git clone --depth=1 "$REPO_TO_LINT" .
+      
+      - name: lint
+        run: npx awesome-lint
```
Comment (aureooms) [inline .github/workflows/main.yml:43]: How does github now that it needs to install `awesome-lint`?

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - uses: actions/checkout@v2
+
+      # by default github actions doesn't fetch the whole repo so we are getting the latest commit on main
+      - name: fetch
+        run: git fetch origin main:main --depth=1
+
+      # find the repo in the git diff and then set it to an env variables
+      - name: find repo
+        run: git diff main readme.md | grep ^+ | grep -Eo 'https.*#readme' | sed 's/#readme//' | xargs -I {} echo "::set-env name=REPO_TO_LINT::{}"
```
Comment (aureooms) [inline .github/workflows/main.yml:32]: Should it not check that the repo is on GitHub (and maybe of the form `https://github.com/[^/]*/[^/]*#readme`)?

**Review by aureooms — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml (hunk 1 above)
Comment (sindresorhus) [reply .github/workflows/main.yml:43]: @aureooms Instead of asking questions, you could easily Google how the `npx` command works.

**Comment by sindresorhus** [comment]
> If readme is changed and no new URL is detected in the readme it will fail with a helpful error message.

I think it should just stay silent then. There are valid PRs that are not list additions.

**Comment by sindresorhus** [comment]
I think this is a good idea. Sorry for not getting to this sooner.

**Comment by sindresorhus** [comment]
@RayBB Friendly bump :)

**Comment by RayBB** [comment]
@sindresorhus I had to make a new PR because old fork was deleted.

New PR address your comment #1962

**Comment by Vikramrama** [comment]
> I really appreciate this project and wanted to try to give back so it's a little easier for everyone involved.
> To that end, I created this PR to **automatically lint new lists** when someone opens a PR to add them.
> 
> ## Benefits
>   - It will save time of maintainers by allowing them to look at the error message first
>   - It will provide immediate feedback to people who make a PR
>   - It helps people having trouble installing `awesome-lint`
>   - It is free
> 
> ## Risks
>   - As far as I know, there is no risk to users or maintainers unless.
> 
> ## What it does
> - PR will lint **NEW** lists that are requesting to be added.
> - PR will **NOT** lint [this](https://github.com/sindresorhus/awesome/) repo itself. If that functionality is desired it can happen in a separate PR (see https://github.com/sindresorhus/awesome/pull/1794).
> - PR will **NOT** lint the current lists in the readme.
> 
> ## How it does it
> This PR creates a GitHub action that will do the following every time a new commit is pushed to an open PR:
> 1. Find the URL that is being added using `git diff`
> 2. Clone that repo
> 3. Run `awesome-lint`on that repo.
> 
> The above 3 steps will **ONLY** run if `readme.md` is changed.
> If readme is changed and no new URL is detected in the readme it will fail with a helpful error message.
> Error can be ignored if it is a cosmetic PR
> 
> [Here](https://github.com/RayBB/awesome/pull/1) are examples of failing commits.
> 
> ## Final notes
> 
> "Squash and Merge" this PR or let me know and I will squash it manually.
> 
> Feel free to edit this any way you see fit or ask me to make the edits. 👍
> If there are any questions on how this works drop them here and I'll try my best to answer.
> 
> Again, thanks so much for all the hard work the maintainers put into this repo! 👏 

21f0a11fd59a12e08884f2ec0ef88d1cad86c78d

**Pull request closed without merging**